tests/optimisation_test/recipe_converter.py

```diff
@@ -3,58 +3,84 @@
+            for key, value in recipe_dict.items():
+                if isinstance(value, str) and re.search(r"^{{float+.+?all}}$", value):
+                    self.placeholder_keys_all.append(key)
+                    self.placeholders.append(value)
+                elif isinstance(value, str) and re.search(r"^{{float+.+?any}}$", value):
+                    self.placeholder_keys_any.append(key)
                     self.placeholders.append(value)
                 else:
                     self.find_placeholders(value)
-        elif isinstance(self._data, list):
-            for item in self._data:
+        elif isinstance(recipe_dict, list):
+            for item in recipe_dict:
                 self.find_placeholders(item)
+        return self.placeholder_keys_any, self.placeholder_keys_all
              
-    def update_values(self,data, *args):
-        self.data = data
-        _output_pd = args[0]
-        _shape = _output_pd.shape
-        _total_columns = _shape[1]
+    def update_values(self, recipe_dict, _type, param_dict):
+        _keys_optimised_parameters_dict = list(param_dict.keys())
+        if _type == 'all':
+           self.placeholder_keys = self.placeholder_keys_all
+        elif _type == 'any':
+            self.placeholder_keys = self.placeholder_keys_any
+        else:
+            logging.error('improper placeholder type')
         for _key in self.placeholder_keys:
-            for col in range(_total_columns):
-                _head = _output_pd.columns[col]
-                if _key == _head:
-                    _values = _output_pd[_head].tolist()
-                    print(_key,_values)
-                    _data_update = nested_update(self.data, key = str(_key), value = _values)
-                    self.data = _data_update
-                elif _key != _head:
-                    pass      
-
-    def create_optimized_recipe(self, path):
+            for col in range(len(param_dict)):
+                if _key == _keys_optimised_parameters_dict[col]:
+                    if _type == 'all':
+                        _value = list(param_dict[_keys_optimised_parameters_dict[col]].values())
+                    elif _type == 'any':
+                        pass
+                        _value = float(param_dict[_keys_optimised_parameters_dict[col]][0])
+                    else:
+                        logging.error('improper placeholder type')
+                    print(_value)
+                    _data_update = nested_update(recipe_dict, key = str(_key), value = _value)
+                    recipe_dict = _data_update
+                elif _key != _keys_optimised_parameters_dict[col]:
+                    pass
+        return recipe_dict    
+
+    def generate_recipe(self, _recipe_dict, path):
         with open(path, 'w') as file:
-            print(self.data)
-            yaml.dump(self.data, file)
+            print(_recipe_dict)
+            yaml.dump(_recipe_dict, file)
```

Comment:
I suggest changing this method to be generate_recipe(pd.DataFrame) that will take an input DataFrame and replace all the relevant values in their correct place in the template_dict.
The file writing bit I would add it as an internal function called _write_recipe()

Reply:
done